MyBatis fails to find constructor for records

# Issue
While upgrading an application from `SpringBoot2/MyBatisSpring2` to `SpringBoot3/MyBatisSpring3` I noticed errors while running the tests:
```
Failed to find a constructor in 'com.example.demo.Foo' by arg names [id, name]. There might be more info in debug log.
```
Debugging `org.apache.ibatis.mapping.ResultMap` showed that the `argNames` were resolved correctly yet `argTypesMatch()` returned false, which in turn lead to the error above.

# Workaround
* use `class` instead of `record`
* use `javaType` in the `resultMap`
* explicit downgrade (via `pom.xml` to `3.5.9`)



## MyBatis version
3.5.11

## Database vendor and version
PostgreSQL 14.4

## Test case or example project
https://github.com/tzie0062/issue-mybatis-record

## Steps to reproduce
Using a resultMap with records

## Expected result
Mapper works

## Actual result
Mapper fails with
```
Caused by: org.apache.ibatis.builder.BuilderException: Error in result map 'com.example.demo.FooRepository.foo_map'. Failed to find a constructor in 'com.example.demo.Foo' by arg names [id, name]. There might be more info in debug log.
	at org.apache.ibatis.mapping.ResultMap$Builder.build(ResultMap.java:130)
	at org.apache.ibatis.builder.MapperBuilderAssistant.addResultMap(MapperBuilderAssistant.java:208)
	at org.apache.ibatis.builder.ResultMapResolver.resolve(ResultMapResolver.java:47)
	at org.apache.ibatis.builder.xml.XMLMapperBuilder.resultMapElement(XMLMapperBuilder.java:289)
	at org.apache.ibatis.builder.xml.XMLMapperBuilder.resultMapElement(XMLMapperBuilder.java:254)
	at org.apache.ibatis.builder.xml.XMLMapperBuilder.resultMapElements(XMLMapperBuilder.java:246)
	at org.apache.ibatis.builder.xml.XMLMapperBuilder.configurationElement(XMLMapperBuilder.java:119)
	... 102 more
```

--- a/src/main/java/org/apache/ibatis/builder/MapperBuilderAssistant.java
+++ b/src/main/java/org/apache/ibatis/builder/MapperBuilderAssistant.java
@@ -437,7 +437,7 @@ public class MapperBuilderAssistant extends BaseBuilder {
     if (javaType == null && property != null) {
       try {
         MetaClass metaResultType = MetaClass.forClass(resultType, configuration.getReflectorFactory());
-        javaType = metaResultType.getSetterType(property);
+        javaType = metaResultType.getGetterType(property);
       } catch (Exception e) {
         // ignore, following null check statement will deal with the situation
       }

--- a/src/main/java/org/apache/ibatis/mapping/ResultMap.java
+++ b/src/main/java/org/apache/ibatis/mapping/ResultMap.java
@@ -128,8 +128,8 @@ public class ResultMap {
         final List<String> actualArgNames = argNamesOfMatchingConstructor(constructorArgNames);
         if (actualArgNames == null) {
           throw new BuilderException("Error in result map '" + resultMap.id + "'. Failed to find a constructor in '"
-              + resultMap.getType().getName() + "' by arg names " + constructorArgNames
-              + ". There might be more info in debug log.");
+              + resultMap.getType().getName() + "' with arg names " + constructorArgNames
+              + ". Note that 'javaType' is required when there is no readable property with the same name ('name' is optional, BTW). There might be more info in debug log.");
         }
         resultMap.constructorResultMappings.sort((o1, o2) -> {
           int paramIdx1 = actualArgNames.indexOf(o1.getProperty());

--- a/src/test/java/org/apache/ibatis/submitted/record_type/RecordTypeMapper.java
+++ b/src/test/java/org/apache/ibatis/submitted/record_type/RecordTypeMapper.java
@@ -32,6 +32,12 @@ public interface RecordTypeMapper {
   @Select("select val, id, url from prop where id = #{id}")
   Property selectProperty(int id);
 
+  @Arg(name = "id", column = "id", id = true)
+  @Arg(name = "value", column = "val")
+  @Arg(name = "URL", column = "url")
+  @Select("select val, id, url from prop where id = #{id}")
+  Property selectPropertyNoJavaType(int id);
+
   @Insert("insert into prop (id, val, url) values (#{id}, #{value}, #{URL})")
   int insertProperty(Property property);
 

--- a/src/test/java/org/apache/ibatis/submitted/record_type/RecordTypeTest.java
+++ b/src/test/java/org/apache/ibatis/submitted/record_type/RecordTypeTest.java
@@ -53,6 +53,16 @@ class RecordTypeTest {
   }
 
   @Test
+  void shouldResolveConstructorArgType() {
+    try (SqlSession sqlSession = sqlSessionFactory.openSession()) {
+      RecordTypeMapper mapper = sqlSession.getMapper(RecordTypeMapper.class);
+      Property prop = mapper.selectPropertyNoJavaType(1);
+      assertEquals("Val1!", prop.value());
+      assertEquals("https://www.google.com", prop.URL());
+    }
+  }
+
+  @Test
   void testSelectRecordAutomapping() {
     try (SqlSession sqlSession = sqlSessionFactory.openSession()) {
       RecordTypeMapper mapper = sqlSession.getMapper(RecordTypeMapper.class);

